src/java/org/apache/cassandra/index/sai/disk/v2/V2VectorIndexSearcher.java

```diff
@@ -369,7 +371,13 @@ public RangeIterator limitToTopResults(QueryContext context, List<PrimaryKey> ke
             for (int i = 0; i < keysInRange.size(); i++)
             {
                 var primaryKey = keysInRange.get(i);
-                long sstableRowId = primaryKeyMap.exactRowIdOrInvertedCeiling(primaryKey);
+
+                // turn the pk back into a row id, with a fast path for the case where the pk is from this sstable
+                long sstableRowId = -1;
+                assert primaryKey instanceof PrimaryKeyWithSource : "Expected PrimaryKeyWithSource, got " + primaryKey;
+                var pkws = (PrimaryKeyWithSource) primaryKey;
+                if (pkws.getSourceSstableId() == sstableId)
```

Comment:
Nit: we have access to sstableId from `primaryKeyMap.getSSTableId()`, which saves us from needing to change the scope of `indexDescriptor`.

Also, I assume the reference equality is sufficient here because of the way we create these PKWS objects, but just in case, I think we should use `.equals`, which does a reference check in all of the implementations, so it shouldn't introduce much overhead.